receiver/prometheusexecreceiver/receiver.go

```diff
@@ -0,0 +1,313 @@
+	scrapeConfig.HonorLabels = false
+	// This is a default Prometheus scrape config value, which indicates that timestamps of the scrape should be respected
+	scrapeConfig.HonorTimestamps = true
+
+	// Set the proper target by creating one target inside a single target group (this is how Prometheus wants its scrape config)
+	scrapeConfig.ServiceDiscoveryConfig = sdconfig.ServiceDiscoveryConfig{
+		StaticConfigs: []*targetgroup.Group{
+			{
+				Targets: []model.LabelSet{
+					{model.AddressLabel: model.LabelValue(fmt.Sprintf("localhost:%v", cfg.Port))},
+				},
+			},
+		},
+	}
+
+	return &prometheusreceiver.Config{
+		PrometheusConfig: &config.Config{
+			ScrapeConfigs: []*config.ScrapeConfig{scrapeConfig},
+		},
+	}
+}
+
+// getSubprocessConfig returns the subprocess config after the correct logic is made
+func getSubprocessConfig(cfg *Config, customName string) (*subprocessmanager.SubprocessConfig, error) {
+	if cfg.SubprocessConfig.Command == "" {
+		return nil, fmt.Errorf("no command to execute entered in config file for %v", cfg.Name())
+	}
+
+	subprocessConfig := &subprocessmanager.SubprocessConfig{}
+
+	subprocessConfig.Command = cfg.SubprocessConfig.Command
+	subprocessConfig.Env = cfg.SubprocessConfig.Env
+
+	return subprocessConfig, nil
+}
+
+// getCustomName will return the receiver's given custom name or try to generate one if none was given
+func getCustomName(cfg *Config) string {
+	// Try to get a custom name from the config (receivers should be named prometheus_exec/customName)
+	splitName := strings.SplitN(cfg.Name(), "/", 2)
+	if len(splitName) > 1 && splitName[1] != "" {
+		return splitName[1]
+	}
+	// fall back to the first part of the string, prometheus_exec, which should only happen if there is a single prometheus_exec receiver configured
+	return splitName[0]
+}
+
+// manageProcess will put the process in an infinite starting loop which goesd through the following steps
+// If the port is not defined by the user, one is generated and a new metrics receiver is built with the new port
+// All instances of {{port}} are replaced with the actual port (either defined by the user or generated)
+// Start the receiver if it is stopped (either first iteration or if it was previously shutdown)
+// We then start the subprocess, get its runtime and decide if the process is considered healthy or not -> the receiver is shutdown if it was deemed unhealthy
+// Finally, the wait time before the subprocess is restarted is computed and this goroutine sleeps for that amount of time, before restarting the loop from the start
+func (wrapper *prometheusReceiverWrapper) manageProcess() {
+	var (
+		elapsed       time.Duration
+		newPort       int
+		crashCount    int
+		err           error
+		subprocessErr error
+	)
+
+	for {
+
+		// Generate a port if none was specified and if process is unhealthy (Receiver has been stopped)
+		generatePort := wrapper.originalPort == 0 && elapsed <= healthyProcessTime
```

Comment:
c/f with line 194 where you have a slightly different definition of unhealthy (also this logic should not be duplicated)

Reply:
They're different! This line you commented on decides whether or not to generate a new port for the process (user had not defined a port and process has not had a long runtime).

The other line checks whether to start the receiver or not (first run or unhealthy or new port was generated). 

I'll update the comments to make more sense. I see what you mean around the confusion of "healthy".

Reply:
Why do we generate a new port if the process has had a long runtime but not otherwise?

It would be nice if you could make your logic always listen for a port (whether that port be 0 or a user specified number) and run all the same code (update placeholders, etc) regardless. That would be very easy to follow.

Reply:
If the process has had a long runtime, we want to give it the chance to restart on the same port to not lose any metrics that may have some kind of state. Restarting on another port implies shutting down the receiver and restarting it which is more destructive!